BuildTools/ModFile/BuildProperties.cs

```diff
@@ -0,0 +1,258 @@
+﻿using Microsoft.Build.Framework;
+using System;
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using System.Text.RegularExpressions;
+
+namespace tModLoader.BuildTools.ModFile;
+
+public class BuildProperties
+{
+	private List<string> _dllReferences = new();
+	private List<ModReference> _modReferences = new();
+	private List<ModReference> _weakReferences = new();
+	private string[] _sortAfter = Array.Empty<string>();
+	private string[] _sortBefore = Array.Empty<string>();
+	private string[] _buildIgnores = Array.Empty<string>();
+	private string _author = "";
+	public Version Version = new(1, 0);
+	private string _displayName = "";
+	private bool _noCompile = false;
+	private bool _hideCode = false;
+	private bool _hideResources = false;
+	public bool IncludeSource = false;
+
+	private string _eacPath = "";
+
+	// This .tmod was built against a beta release, preventing publishing.
+	internal bool Beta = false;
+	private string _homepage = "";
+	private string _description = "";
+	private ModSide _side;
+
+	public IEnumerable<ModReference> Refs(bool includeWeak) =>
+		includeWeak ? _modReferences.Concat(_weakReferences) : _modReferences;
+
+	public IEnumerable<string> RefNames(bool includeWeak) => Refs(includeWeak).Select(dep => dep.mod);
+
+	private static IEnumerable<string> ReadList(string value)
+		=> value.Split(',').Select(s => s.Trim()).Where(s => s.Length > 0);
+
+	private static void WriteList<T>(IEnumerable<T> list, BinaryWriter writer) {
+		foreach (var item in list)
+			writer.Write(item!.ToString()!);
+
+		writer.Write("");
+	}
+
+	internal static BuildProperties ReadTaskItems(IEnumerable<ITaskItem> taskItems) {
+		BuildProperties properties = new BuildProperties();
+
+		foreach (ITaskItem property in taskItems) {
+			string propertyName = property.ItemSpec;
+			string propertyValue = property.GetMetadata("Value");
+
+			// Make the first letter lowercase
+			propertyName = char.ToLowerInvariant(propertyName[0]) + propertyName.Substring(1);
+
+			ProcessProperty(properties, propertyName, propertyValue);
+		}
+
+		var refs = properties.RefNames(true).ToList();
+		if (refs.Count != refs.Distinct().Count())
+			throw new Exception("Duplicate mod/weak reference");
```

Comment:
If you have duplicate mod references, it will not throw an error right now as MSBuild either merges the references, or only uses one of the `ModReference`